### ISSUE TITLE
Add support for using source-based analyser instead of snapshot

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#266]
This would make it easier to use Dart Code for testing tweaks to the Analyzer without having to build new snapshots. This would be useful for devs/contributors to the analyzer.

@scheglov I'm not sure how to launch the Analyzer using source instead of snapshot, it is it just a case of changing the filename passed as the first arg to the VM?

So, if I add a new option that lets you specify the path:

```js
    "dart.analyzerPath": "M:\something\somefile.dart"
```

And then swap that path in place of the `.snapshot` file when launching the analyzer?

**Comment by scheglov** [comment]
Yes, you just specify the path of the file that you want to run instead of the snapshot from SDK.

I run Analysis Server using code like this: `/Users/scheglov/Applications/dart-sdk/bin/dart /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analysis_server/bin/server.dart`.

**Comment by DanTup** [comment]
@scheglov I've added an option for this, however when testing it I get an error trying to run it:

```text
PS M:\Coding> dart M:\Coding\Applications\sdk\pkg\analysis_server\bin\server.dart
  Unhandled exception:
    Could not import "package:plugin/plugin.dart" from "package:analysis_server/starter.dart":
    FileSystemException: Cannot open file, path = 'M:\Coding\Applications\sdk\third_party\pkg\plugin\lib\plugin.dart'
    (OS Error: The system cannot find the path specified., errno = 3)
```

Is this just because there's some setup required to run the Analyzer from the SDK repo (like building stuff, or fetching dependencies)? My `thirdparty\pkg` folder just has a single file named `unittest.tar.gz.sha1`.

If you think this would normally work, I'm happy to just push this. It's a hidden option so users won't see in the settings.

**Comment by DanTup** [comment]
I included this in 1.4.1, so if you want to try it you can add something like this to your user settings:

```text
  "dart.analyzerPath": "/Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analysis_server/bin/server.dart"
```

There were already settings for the below. If there's anything else that would be useful for tracking down bugs (or using VS Code for analyzer dev!), please let me know.

- `dart.analyzerDiagnosticsPort` - passed to analyzer as `--port=`
- `dart.analyzerObservatoryPort` - passed to Dart VM as `--observe=`
- `dart.analyzerLogFile` - Dart Code log file for communication (truncated)
- `dart.analyzerInstrumentationLogFile` - passed to analyzer as `--instrumentation-log-file=`
- `dart.analyzerAdditionalArgs` - appended to the end of the other analyzer args (eg. was used previously to pas the new-driver flag)
- `dart.sdkPath` - Path to the Dart SDK (otherwise it's used from PATH)

All of these settings work both in global user settings and workspace settings (so you can set them for a specific project) and changing many of them will prompt you to reload the project to re-launch the analyzer.

**Comment by DanTup** [comment]
I finally got around to testing/using this.. It seems to work great :-)

Not only that; but because VS Code supports per-workspace settings, it's easy to have one Dart Code window loading a sample project that points to the source-based analyzer whilst actually working on that analyzer in another window which is using the snapshot-based analyzer.
